internal/pkg/skuba/addons/cilium.go

```diff
@@ -58,10 +62,27 @@ func renderCiliumTemplate(addonConfiguration AddonConfiguration) string {
 	return ciliumManifest
 }
 
+type preflightContext struct {
+	CiliumImage string
+}
+
+func renderCiliumPreflightTemplate(addonConfiguration AddonConfiguration) (string, error) {
+	template, err := template.New("").Parse(ciliumPreflightManifest)
```

Comment:
@ereslibre Unfortunately I have no ideas for adding the Cilium preflight (a temporary deployment just for performing the etcd->CRD migration) into addon template logic. I'm ooen for suggestions and I will try to come up with some better solution.

Reply:
Addons have the concept of callbacks that you can implement in the cilium one:

- `beforeApply`: https://github.com/SUSE/skuba/blob/abe9ca15bc09939288538b1b43dcf9375b865beb/internal/pkg/skuba/addons/addons.go#L235
- `afterApply: https://github.com/SUSE/skuba/blob/abe9ca15bc09939288538b1b43dcf9375b865beb/internal/pkg/skuba/addons/addons.go#L261

When you register the addon at https://github.com/SUSE/skuba/blob/master/internal/pkg/skuba/addons/cilium.go#L30, you can set these callbacks.

Would that help?

Reply:
Most probably yes, that will help. Thanks!

Reply:
Or no... rather not, sorry.

I'm already using `beforeApply` callback and the thing is that I implemented the whole logic there, including generation of template, which is almost copied&pasted from the generic code for deployments in `addons`.

https://github.com/SUSE/skuba/blob/ee9b9a38ea30ccfce4c0c25783619757a9ca9e5d/internal/pkg/skuba/addons/cilium.go#L69-L80

https://github.com/SUSE/skuba/blob/ee9b9a38ea30ccfce4c0c25783619757a9ca9e5d/internal/pkg/skuba/addons/addons.go#L157-L171

So I was rather thinking about some way to make deployment of "precheck" Cilium generic in the similar way as for normal addon deployments, but I didn't come up with any idea.

Or is the way that I currently implemented that fine for you?

Reply:
Let me focus on reviewing this properly, will give my 2 cents later today.

Reply:
Let me know if I understand this correctly: to do the upgrade, firstly cilium-preflight daemonset must be deployed. Then, wait until the number of READY pods from the cilium-preflight daemonset is the same as the number of Cilium pods running. After that, we can delete the cilium-preflight daemonset and proceed with the upgrade.

The only way to deploy right now is by registering the Addon into the Addons map and providing the manifest. At some point, the Addon is processed and the "kubectl apply" is executed pointing to the manifests that we registered. However, there are two problems:
1 - This mechanism does not include a way to "wait until the number of READY pods from the cilium-preflight daemonset is the same as the number of Cilium pods running". It just deploys what is in the manifest, no extra logic.
2 - After the cilium-preflight step is done, we must do some stuff related to the upgrade like MigrateEtcdToCrd before the real deployment kicks-in. This requires calls to the kube-API and can't be expressed with a manifest

We could do all this in the beforeApply callback but then we would need to use "kubectl apply" there to deploy the cilium-preflight daemonset. Then check for it to be ready, do the related stuff like MigrateEtcdToCrd and then delegate the real deployment to Addons. However, that would look like a hack.